NBitcoin/Socks/SocksHelper.cs

```diff
@@ -66,17 +76,57 @@ internal static byte[] CreateConnectMessage(EndPoint endpoint)
 			return sendBuffer;
 		}
 
-		public static async Task Handshake(Socket socket, EndPoint endpoint, CancellationToken cancellationToken)
+		public static async Task Handshake(Socket socket, EndPoint endpoint, bool changeTorIdentities, CancellationToken cancellationToken)
 		{
 			NetworkStream stream = new NetworkStream(socket, false);
-			await stream.WriteAsync(SelectionMessage, 0, SelectionMessage.Length).WithCancellation(cancellationToken).ConfigureAwait(false);
+			var selectionMessage = changeTorIdentities ? SelectionMessageUsernamePassword : SelectionMessageNoAuthenticationRequired;
+			await stream.WriteAsync(selectionMessage, 0, selectionMessage.Length).WithCancellation(cancellationToken).ConfigureAwait(false);
 			await stream.FlushAsync().WithCancellation(cancellationToken).ConfigureAwait(false);
 
 			var selectionResponse = new byte[2];
 			await stream.ReadAsync(selectionResponse, 0, 2).WithCancellation(cancellationToken).ConfigureAwait(false);
 			if (selectionResponse[0] != 5)
 				throw new SocksException("Invalid version in selection reply");
-			if (selectionResponse[1] != 0)
+			if (selectionResponse[1] == 2)
+			{
+				// https://tools.ietf.org/html/rfc1929#section-2
+				// Once the SOCKS V5 server has started, and the client has selected the
+				// Username / Password Authentication protocol, the Username / Password
+				// subnegotiation begins.  This begins with the client producing a
+				// Username / Password request:
+
+				// Create a random string 21 char string.
+				const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+				var identity = new string(Enumerable.Repeat(chars, 21)
+				  .Select(s => s[new Random().Next(s.Length)]).ToArray());
+
+				var username = identity;
+				var password = identity;
+				var uName = Encoding.UTF8.GetBytes(username);
+				var passwd = Encoding.UTF8.GetBytes(password);				
```

Comment:
are you sure about this? UTF8 by default insert a BOM marker as first byte.

Reply:
BOM is inserted as first char in a text when you do `Encoding.UTF8.GetString` but here we are getting the bytes.

Reply:
Yes, I'm sure. Both in my `PasswdField` and `UNameField` I have this comment in the code: `// Tor accepts UTF8 encoded passwd`

I have this comment in the code: https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/TorSocks5/Models/Fields/ByteArrayFields/PasswdField.cs#L13